safekeeper/src/debug_dump.rs

```diff
@@ -53,15 +54,75 @@ pub struct Args {
 }
 
 /// Response for debug dump request.
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, Serialize)]
```

Comment:
I forgot about `Deserialize` here when I suggested `TimelineDumpSer`...
We need to have `Deserialize` for `pull_timeline` to work properly. The easiest way to do this is probably to copy `Response` struct with a `Timeline` instead of `TimelineDumpSer`, so it can be used for deserialization when we want to query debug_dump on another safekeeper in pull_timeline.

Reply:
I have question. @petuhovskiy 
If we should keep the `debug_dump` query in `pull_timeline`, I think we also should keep the `timelines: Vec<Timeline>` inside the `Response` struct. (simply, keep the `Timeline`, not introducing the `TimelineDumpSer`)
The reason is that we cannot implement the `Deserialize` matching the `Seriailze`, which does streaming of `Timeline` info.
In this case(we keep the `Timeline`), we have no JSON build streaming, but just http response streaming.
And the `debug_dump` query would still need much resource(RAM).
Am I right?

Reply:
> And the debug_dump query would still need much resource(RAM).

Not really, `pull_timeline` requests `debug_dump` only for a single timeline. In this case there are no issues with RAM usage.

But we also have another case, when we get `debug_dump` for all existing timelines, for further inspection. In this case RAM is important, because we can have 100k+ timelines and each dump can have a size 10KB+ in JSON.

Reply:
Yes, you are right.
I am sure it needs much resource in worst case(dumping all timelines).

Reply:
> Should I keep Timeline, instead of TimelineDumpSer?

Use `TimelineDumpSer` in `Response` inside `debug_dump.rs`. It's bad for any HTTP request to use much more than 1MB of RAM, so streaming response is always preferred.

> If you believe the use of debug_dump query in pull_timeline is priority, we go this way.
> #
> Then, we should find/implement a way to pull timeline info from safekeepers, without using debug_dump query.

`pull_timeline` is important and it can be implemented without `debug_dump`, but for now it's easier to continue using everything as is.

Just copy `Response` to `pull_timeline.rs` and use `Timeline` instead of `TimelineDumpSer` in that copy. `pull_timeline.rs` will be able to fetch and deserialize response into its own copy, `debug_dump.rs` will have its own copy for serialization.

It's probably also possible to use generics as `Response<TimelineDumpSer or Timeline>`, but it's very confusing, I don't like it.

Reply:
@petuhovskiy 
Done https://github.com/neondatabase/neon/pull/5731/commits/e5d9eabb0c02e19d50ab7d18e2bd9d9347e84674

Reply:
I would like you to do review again.
On the other hand, I will try to test the improvements.